crawler4j-core/src/main/java/edu/uci/ics/crawler4j/crawler/WebCrawler.java

```diff
@@ -405,184 +383,187 @@ public void visit(Page page) {
-                        } else {
-                            webURL.setDocid(-1);
-                            webURL.setDepth((short) (curURL.getDepth() + 1));
-                            if ((maxCrawlDepth == -1) || (curURL.getDepth() < maxCrawlDepth)) {
-                                if (shouldVisit(page, webURL)) {
-                                    if (robotstxtServer.allows(webURL)) {
-                                        webURL.setDocid(docIdServer.getNewDocID(webURL.getURL()));
-                                        toSchedule.add(webURL);
-                                    } else {
-                                        logger.debug(
-                                                "Not visiting: {} as per the server's \"robots.txt\" " +
-                                                        "policy", webURL.getURL());
-                                    }
-                                } else {
-                                    logger.debug(
-                                            "Not visiting: {} as per your \"shouldVisit\" policy",
-                                            webURL.getURL());
-                                }
-                            }
-                        }
-                    }
-                    frontier.scheduleAll(toSchedule);
-                } else {
-                    logger.debug("Not looking for links in page {}, "
-                                    + "as per your \"shouldFollowLinksInPage\" policy",
-                            page.getWebURL().getURL());
-                }
-
-                boolean noIndex = myController.getConfig().isRespectNoIndex() &&
-                        page.getContentType() != null &&
-                        page.getContentType().contains("html") &&
-                        ((HtmlParseData) page.getParseData())
-                                .getMetaTagValue("robots").
-                                contains("noindex");
-
-                if (!noIndex) {
-                    visit(page);
-                }
-            }
-        } catch (PageBiggerThanMaxSizeException e) {
-            onPageBiggerThanMaxSize(curURL.getURL(), e.getPageSize());
-        } catch (ParseException pe) {
-            onParseError(curURL, pe);
-        } catch (ContentFetchException | SocketTimeoutException cfe) {
-            onContentFetchError(curURL);
-            onContentFetchError(page);
-        } catch (NotAllowedContentException nace) {
-            logger.debug(
-                    "Skipping: {} as it contains binary content which you configured not to crawl",
-                    curURL.getURL());
-        } catch (IOException | InterruptedException | URISyntaxException | RuntimeException e) {
-            onUnhandledException(curURL, e);
-        } finally {
-            if (fetchResult != null) {
-                fetchResult.discardContentIfNotConsumed();
-            }
-        }
-    }
+		/**
```

Comment:
@reformatting: I have a custom formatter I use inside all my Eclipses that uses tabs, but lots of people get thought that spaces are better because  you might want to look at code inside vi or something. I no longer argue about this, and try to keep the spaces when they are there, but in this case there were a lot of changes going on and mixing tabs and spaces indentation inside a method is kind of horrible :-).

@pointers:
The parse method is broken up into 4 different parts depending on the content to parse (binary, css, text, html). Different parts "do their thing" in different ways, the "thing" being: parse + set outgoing urls. Depending on the part there was some exception handling, with only the binary exception handling taking into account the "haltOnError" configuration parameter. For the rest most parts do "catch every Exception, log and rethrow a ParseException" (sometimes without keeping the original exception). Later on the WebCrawler catches the ParseException and does nothing, but "allows" to handle it as you like (the logging is already enforced by that point).

What I wanted to achieve was: handle the exceptions depending on the type of content or have a general handling. When the general handling kicks in, the page won't be processed any further (idem as original), this also means the visit()-method won't be called (which I use to store the parsed content). To handle the exceptions depending on the type of content, it is now a lot easier to do this inside the XxxParseData-classes because we have the create()-methods that allow overriding the specific implementation.

In the end (next commit) I want to get to the more interesting point where you can express "I don't care about errors in parsing binary or css content" (i.e. no "haltOnError"), it just means that the outgoing urls are not extracted (which you would typically log as a warning), but I rather store a possible wrong css file than lose all the css content because of a parsing issue.

Bonus: it is possible to avoid the general exception handling, but still stop the further processing (calling the visit()-method) by throwing a NotAllowedContentException (idem as original).

Future enhancement:
As I mentioned, parsing the 4 different content parts and getting the outgoing urls is not always initiated in the same way. Initiating it by calling a setter is also very confusing, so I would rename for example `parseData.setOutgoingUrls(page.getWebURL());` later on.